R/mlflow/R/model.R

```diff
@@ -0,0 +1,121 @@
+#' Save Model for MLflow
+#'
+#' Saves model in MLflow's format that can later be used
+#' for prediction and serving.
+#'
+#' @param fn The serving function that will perform a prediction.
+#' @param path Destination path where this MLflow compatible model
+#'   will be saved.
+#'
+#' @importFrom yaml write_yaml
+#' @export
+mlflow_save_model <- function(fn, path = "model") {
```

Comment:
This diverts a bit from how it is done in python. 

In python we have Model class which defines the structure of MLmodel file with all shared fields (e.g. time_created and runUuid and the flavors) and model flavors which provide implementation of different types of models. Each flavor provides implementation of save_model and load_model.

I think it would be great if we could keep the same structure so that people can add other flavors in a similar way. Can we have a Model.R with functionality related to model logging and MLmodel file and Rfunction.R with a Rfunction model implementation?

For the flavor side:
* save_model should take an optional Model object (or some way to be passed at least runUuid).
* save_model should take an optional dependencies ( r-dependencies.txt and conda.yaml )
* load model should take optional runUuid and if provided load the model from an artifact. 
* it would be great if I could have a generic flavor with similar capability to PyFunc. 

The current Rfunction is great but it seems a bit less generic than what we have in python. One key difference is that in python we let users to pass in a function to load AND initialize the model. I think it would be great if we could have a Rfunc which would first call user provided init function with the path to the model as an argument. In python this function is provided as a source code (python module) which I think is nice because it means people can genrate pyfunc models without executing any python and I think this would make sense in R as well. That way all models that are also supported in R could easily be imported as an Rfunc. For example, it would be great if it would be possible to directly generate an R func flavor with my keras model stored in python without the need to have R installed.

Reply:
For the flavor side feedback:
- Yes, `save_model` now can take a `model`, this was addressed with https://github.com/rstudio/mlflow/pull/12
- Yes, dependencies support was just added directly into our `master` branch
- Yes, implemented last week
- About generic flavor that are based on code. While I agree, this is something that we will have to pause for now. Worth mentioning `RFunc` is implemented as generic as `PyFunc`; however, it does require you to package the model and init function directly in R. Now, that said, the PR I mentioned made the R models generic, so a package could define an `r_code` flavor that somehow converts code to a function, etc. So, at least, users motivated to get this working should be able to get something in place with not much work.